### DISCUSSION TITLE
Is it possible to develop ANEs with the C#?

### DISCUSSION THREAD

**Discussion by shaucky** [airsdk/Adobe-Runtime-Support#2680] [Q&A]
Hey guys,
 
This may seem like a very low level question to you… Just, I'm  very disconnected from the AIR/AS3 developer community, much of what I know about AIR/AS3 comes from Adobe's document and my own experimentation. So, wishing this question won't let u feel strange…
 
I had read referances from https://help.adobe.com/en_US/air/extensions/index.html , it seems Adobe only offer C and Java language version ANE API source code (And i haven't find where the code be offered…).
 
So pls forgive my ignorance… Is it possible to develop ANEs with C Sharp? At first i just thought to develop native program with C#, and call 'em with `NativeProcess`. But then i thought, if there's cross-platform advantages, why don't i take it?
 
I know that frameworks like Xamarin can develop cross-platform program with C#. But can it develop ANEs?
 
ActionScript now has low market share in my country, so i have to try other languages in case i "Unemployment upon graduation".
 
Thank you all.

**Comment by ajwfrost** [comment]
If you want to write a library in C# to implement an ANE, you'd also need a thin wrapper library that can be loaded by AIR and exposes the C-style functions that are required in the documentation and listed in the extension XML file. So it's possible, but not directly..

See e.g. https://stackoverflow.com/questions/2425212/exporting-functions-from-a-c-sharp-class-library

**Reply by shaucky** [reply]
Thanks.
 
It seems that means i can implement specific functions through C#, but in different OS i still must to code specific calling routine with native language (C & Java)? What i want to do is develop background process with C# (on desktop it's easy to understand, on mobile it like Android Service), for example, the AIR/AS3 is mainly responsible for the control of the GUI, and the C# ANE handle relatively complex operations.

**Reply by ajwfrost** [reply]
AIR isn't able to directly load a C# class library, so basically you can create a library using C# but then you would need to create another library in C/C++ that acts as an intermediary. This middle library needs to expose C-style functions as a standard DLL that AIR would then connect to; it needs to load the CLR and the C# library and then access the classes/APIs in the C# library in order to provide the functionality.

So yes, then AIR would have the UI and can call down through this intermediate ANE-style DLL and that DLL would then have the capabilities to call into your C# library. It's all possible, just not particularly straightforward!

thanks

**Reply by itlancer** [reply]
@ajwfrost 
I think it could be great feature to have native support C# and Swift for AIR native extensions.
It will solve a lot of headaches.
Related feature request: https://github.com/airsdk/Adobe-Runtime-Support/issues/1632

**Reply by shaucky** [reply]
Thank you.
 
This seems to be a solution on the Windows platform, how about Android? In my comment, i also wanna take the cross-platform advantage of C#. If the solution can only use on Windows, seems the result has no different with `NativeProcess` calles…
 
According to Adobe's referance, ANE on Android needs developing with Java, so the only way is in some way to implement the Java program to C# program calls? Or anybody knows are there anyway to develop C# ANE for Android in a more convenient way?

**Reply by ajwfrost** [reply]
I believe on Android, C# would be compiled to dex? which means you would likely need a Java wrapper for this too, and call in to the C# code via reflection?

Might be interesting to have a bit of a play with some of these options (@itlancer and also @marchbold interested in your thoughts..)

**Reply by marchbold** [reply]
What exactly is the goal here, as far as I'm aware to develop with c# you need a version of .net? So on Android you would have to include something like the mono libs or some version of the .net runtime which would further bloat your application?

I think the purpose of ANE's is to expose functionality native to the platform so using the language that is native to that platform seems appropriate. And much more relevant than including additional interpreters / runtimes just for other languages? 

Also I think the differences between the platforms is enough that the code you would end up writing in c# wouldn't be any more cross platform than clean implementations in the native languages. 

If you are looking to implement high performance cross platform algorithms then using cross platform c/c++ is a much better approach where there already is native support on all platforms.

I wouldn't mind being able to produce an extension solely in the newer languages (kotlin for android, swift for ios / mac and  c# for windows)  However I don't find the thin wrappers very hard to create and that they don't limit development at all so it's not something high on my priority list. At worst they increase the time to add an ANE function by a few minutes.

**Reply by marchbold** [reply]
@xiaoChi888 To actually implement an Android ANE currently in C# I'd imagine you would need a java wrapper that would call into NDK compiled lib (c/c++?) which would load your c# / .net / mono lib. 

Personally way too many layers for me and I'd question if you'd get any value out of your cross-platform lib this way.

**Reply by shaucky** [reply]
Yeah, an additional runtime will further bloat the package… I really don't have special consider if it is worth…
 
I'm not sure if it is an appropriate practice, all ideas are just attempts. My original intention was to make a set of logic cross-platform with AIR (but not straight using ActionScript), it would be even better to "Code once, run everywhere".
 
A Java wrapper for C# on Android shouldn't be hard to implement, there's a referance from Microsoft: https://learn.microsoft.com/en-us/xamarin/android/get-started/java-developers , i haven't read it carefully.
 
Indeed, from an efficiency point of view, native languages are really better suited for ANE development…
 
These ideas are well worth my consideration, thank you.

**Comment by FischMich** [comment]
This may be a good entry point -> https://github.com/tuarua/FreSharp
C# wrapper for FlashRuntimeExtensions
Overview of some C# based ANE's -> https://github.com/tuarua

**Reply by marchbold** [reply]
Just a note, those references are in regards to c# on windows using .net and not applicable to Android or other platforms.  

But yes this is a version of a thin wrapper converting from c -> c# on windows.

**Reply by pol2095** [reply]
I think [FreSharp](https://github.com/tuarua/FreSharp) work like this, on windows the dll in C++/CLI (CLR) call a dll in C#, I use it for UWP classes, not usable on android.
